src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -0,0 +1,192 @@
+                    _buffer_max_size, AsyncLogBufferSize);
+}
+
+class AsyncLogMapIterator {
+  AsyncLogBuffer& _logs;
+
+ public:
+  AsyncLogMapIterator(AsyncLogBuffer& logs) :_logs(logs) {}
+  bool do_entry(LogFileOutput* output, uint32_t* counter) {
+    using none = LogTagSetMapping<LogTag::__NO_TAG>;
+
+    if (*counter > 0) {
+      LogDecorations decorations(LogLevel::Warning, none::tagset(), output->decorators());
+      stringStream ss;
+      ss.print(UINT32_FORMAT_W(6) " messages dropped due to async logging", *counter);
+      AsyncLogMessage msg(*output, decorations, ss.as_string(true /*c_heap*/));
+      _logs.push_back(msg);
+      *counter = 0;
+    }
+
+    return true;
+  }
+};
+
+void AsyncLogWriter::perform_IO() {
+  // Use kind of copy-and-swap idiom here.
+  // Empty 'logs' swaps the content with _buffer.
+  // Along with logs destruction, all procceeded messages are deleted.
+  //
+  // The operation 'pop_all()' is done in O(1). All I/O jobs are then performed without
+  // lock protection. This guarantees I/O jobs don't block logsites.
+  AsyncLogBuffer logs;
+  { // critical region
+    AsyncLogLocker ml;
+    AsyncLogMapIterator dropped_counters_iter(logs);
+
+    _buffer.pop_all(&logs);
+    // append meta-messages of dropped counters
+    _stats.iterate(&dropped_counters_iter);
+  }
+
+  LinkedListIterator<AsyncLogMessage> it(logs.head());
+  while (!it.is_empty()) {
+    AsyncLogMessage* e = it.next();
+    char* msg = e->message();
+
+    if (msg != nullptr) {
+      e->output()->write_blocking(e->decorations(), msg);
+      os::free(msg);
+    }
+  }
+}
+
+void AsyncLogWriter::run() {
+  assert(_state == ThreadState::Running, "sanity check");
+
+  while (_state == ThreadState::Running) {
+    _sem.wait();
+    perform_IO();
+  }
+
+  assert(_state == ThreadState::Terminated, "sanity check");
+  perform_IO(); // in case there are some messages left
+}
+
+AsyncLogWriter* AsyncLogWriter::_instance = nullptr;
+
+void AsyncLogWriter::initialize() {
+  if (!LogConfiguration::is_async_mode()) return;
+
+  if (_instance == nullptr) {
+    AsyncLogWriter* self = new AsyncLogWriter();
+
+    if (self->_state == ThreadState::Initialized) {
+      Atomic::release_store_fence(&AsyncLogWriter::_instance, self);
+      // All readers of _instance after the fence see non-NULL.
+      // We make use LogOutputList's RCU counters to ensure all synchronous logsites have completed.
+      // After that, we start AsyncLog Thread and it exclusively takee over all logging I/O.
+      for (LogTagSet* ts = LogTagSet::first(); ts != NULL; ts = ts->next()) {
+        ts->wait_until_no_readers();
+      }
+      self->_state = ThreadState::Running;
+      os::start_thread(self);
+      log_debug(logging, thread)("AsyncLogging starts working.");
+    } else {
+      log_warning(logging, thread)("AsyncLogging failed to launch thread. fall back to synchronous logging.");
```

Comment:
"Async logging failed to create thread. Falling back to synchronous logging."

Shouldn't this be done where the os::create_thread failed though?

Reply:
make sense.  I think I can move "Async logging failed to create thread. Falling back to synchronous logging."